httpx/middleware/auth.py

```diff
@@ -0,0 +1,200 @@
+
+        HA1 = digest(A1)
+        if challenge.algorithm.lower().endswith("-sess"):
+            HA1 = digest(b":".join((HA1, challenge.nonce, cnonce)))
+
+        qop = challenge.qop
+        if not qop:
+            digest_data = [HA1, challenge.nonce, HA2]
+        elif qop == b"auth" or b"auth" in qop.split(b",") or b"auth" in qop.split(b" "):
+            digest_data = [challenge.nonce, nc_value, cnonce, b"auth", HA2]
+        elif qop == b"auth-int":
+            raise NotImplementedError("Digest auth-int support is not yet implemented")
+        else:
+            raise ProtocolError(f'Unexpected qop value "{qop!r}" in digest auth')
+
+        key_digest = b":".join(digest_data)
+
+        format_args = {
+            "username": self.username,
+            "realm": challenge.realm,
+            "nonce": challenge.nonce,
+            "uri": path,
+            "response": digest(b":".join((HA1, key_digest))),
+            "algorithm": challenge.algorithm.encode(),
+        }
+        if challenge.opaque:
+            format_args["opaque"] = challenge.opaque
+        if qop:
+            format_args["qop"] = b"auth"
+            format_args["nc"] = nc_value
+            format_args["cnonce"] = cnonce
+
+        header_value = ", ".join(
+            ['{}="{}"'.format(key, to_str(value)) for key, value in format_args.items()]
+        )
+        return "Digest " + header_value
+
+    def _get_client_nonce(self, nonce_count: int, nonce: bytes) -> bytes:
+        s = str(nonce_count).encode()
+        s += nonce
+        s += time.ctime().encode()
+        s += os.urandom(8)
+
+        return hashlib.sha1(s).hexdigest()[:16].encode()
+
+    def _get_nonce_count(self, nonce: bytes) -> typing.Tuple[int, bytes]:
+        """Returns the number of requests made with the same server provided
+        nonce value along with its 8-digit hex representation."""
+        if nonce != self._previous_nonce:
+            self._nonce_count = 1
+        else:
+            self._nonce_count += 1
+        self._previous_nonce = nonce
+        return self._nonce_count, b"%08x" % self._nonce_count
+
+
+class DigestAuthChallenge:
+    def __init__(
+        self,
+        realm: bytes,
+        nonce: bytes,
+        algorithm: str = None,
+        opaque: typing.Optional[bytes] = None,
+        qop: typing.Optional[bytes] = None,
+    ) -> None:
+        self.realm = realm
+        self.nonce = nonce
+        self.algorithm = algorithm or "MD5"
+        self.opaque = opaque
+        self.qop = qop
+
+    @classmethod
+    def from_header(cls, header: str) -> "DigestAuthChallenge":
+        """Returns a challenge from a Digest WWW-Authenticate header.
+
+        These take the form of:
+        `Digest realm="realm@host.com",qop="auth,auth-int",nonce="abc",opaque="xyz"`
+        """
+        scheme, _, fields = header.partition(" ")
+        assert scheme.lower() == "digest"
+        header_dict: typing.Dict[str, str] = {}
+        for field in fields.split(","):
```

Comment:
This doesn't look right on first sight. Eg. `Digest realm="realm@host.com",qop="auth,auth-int",nonce="abc",opaque="xyz"` contains commas inside the quotes which looks like it'd break the split.

Reply:
Good point @tomchristie. Maybe we should be looking at using regexes at this point?

Reply:
Best prior art to consider would be:

* What does requests do here?
* What does werkzeug do here?

Reply:
Looks like they're using the [same helper function](https://github.com/psf/requests/blob/3d968ff34dd9945f2a86396001408ba69c1ab3fd/requests/utils.py#L349)

Reply:
@yeraydiazdiaz Actually it seems to me they use a urllib helper, see [here](https://github.com/psf/requests/blob/3d968ff34dd9945f2a86396001408ba69c1ab3fd/requests/compat.py#L59) and [here](https://github.com/psf/requests/blob/3d968ff34dd9945f2a86396001408ba69c1ab3fd/requests/utils.py#L372).

Reply:
Fabulous, that's a great pointer.